src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2062,444 @@ ShouldInitiateMetadataSync(bool *lockFailure)
 	*lockFailure = false;
 	return shouldSyncMetadata;
 }
+
+
+/*
+ * citus_internal_add_partition_metadata is an internal UDF to
+ * add a row to pg_dist_partition.
+ */
+Datum
+citus_internal_add_partition_metadata(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	char distributionMethod = PG_GETARG_CHAR(1);
+	int colocationId = PG_GETARG_INT32(3);
+	char replicationModel = PG_GETARG_CHAR(4);
+
+	text *distributionColumnText = NULL;
+	char *distributionColumnString = NULL;
+	Var *distributionColumnVar = NULL;
+
+	if (!PG_ARGISNULL(2))
+	{
+		distributionColumnText = PG_GETARG_TEXT_P(2);
+		distributionColumnString = text_to_cstring(distributionColumnText);
+		distributionColumnVar = stringToNode(distributionColumnString);
```

Comment:
We use `stringToNode` which validates the `Var`.  We even have a test:
```SQL
SELECT citus_internal_add_partition_metadata ('test_2'::regclass, 'h', 'non_existing_col', 0, 's');
ERROR:  unrecognized token: "non_existing_col"
```

 The point is that I don't want to change any behavior with this PR, just work as it used to be for the more generic case where the coordinator syncs the metadata. In that case, we use serialized `Var`

Reply:
That test doesn't test everything. It only tests that the input should be a Node that's serialized to a string. If you pass in a stringified version of a `Node` of a different type than `Var` we will still happily cast it to a `Var` without any checking, which can lead to reads past the end of the buffer.

My main point is that it was acceptable when only superuser could do this, but now that regular users can it becomes very dangerous. The easiest way to avoid all these issues seems to be: Give the column as the argument and generate the stringified `Var` based on that inside this function.(it even results in smaller SQL). So essentially call `column_name_to_column` from inside this function, instead of taking it's result as an argument.

Reply:
To be clear, we would still store the stringified `Var` inside `pg_dist_partition`. We would just not allow the user to pass it in directly.

Reply:
> e will still happily cast it to a Var without any checking, which can lead to reads past the end of the buffer.

Wouldn't the following check avoid all sorts of problems? 
```C
	Node *columnNode = stringToNode(columnNodeString);
	if (columnNode == NULL || !IsA(columnNode, Var))
	{
		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
						errmsg("not a valid column")));
	}
```

Reply:
It would, but I'm not certain there's no other ways to cause some bad behaviour by choosing the fields of the `Var` in some weird way. It's much more easy to reason about the safety of us generating the `Var` based on the column name.

Reply:
Also, it's not clear to me from the comments on `stringToNode` in the PG source that it is safe to call on untrusted input.

Reply:
> Also, it's not clear to me from the comments on stringToNode in the PG source that it is safe to call on untrusted input.

That's true, it seems we can ingest wrong Var`s, such as below (e.g., `varcitusno` field):
```C
	SELECT citus_internal_add_partition_metadata ('test_2'::regclass, 'h', '{VAR :varno 1 :varcitusno 1 :vartype 23 :vartypmod -1 :varcollid 0 :varlevelsup 0 :varnosyn 1 :varattnosyn 1 :location 7}', 0, 's');
```

My main motivation for passing the string is that on the coordinator and on the worker, the Var fields might differ, like `varattno` depending on the order of `DROP COLUMN`s or `vartype` depending on the composite type OIDs.

Even though the current master branch behaves wrong on such conditions, I didn't want to change.

I guess given that `stringToNode` expects validated string, I think I'll go with the column name even if we are changing the behavior slightly

Reply:
Maybe I'm missing something, but I don't see how the behaviour changed. In all versions (the one before the PR, the old one in the PR and the one you pushed now), the stringified `Var` is calculated on the worker.

Reply:
Oh, that's true, I was somehow thinking that it is calculated on the coordinator, but that's not true

Reply:
We pass the column name now